calour/io.py

```diff
@@ -286,8 +261,33 @@ def read_open_ms(data_file, sample_metadata_file=None, gnps_file=None, feature_m
     return exp
 
 
+def _read_metadata(ids, f, kwargs):
+    # load the sample/feature metadata file
+    if f is not None:
+        default = {'index_col': 0}
+        if kwargs is None:
+            # use first column as sample ID
+            kwargs = default
+        else:
+            kwargs.update(default)
+        metadata = pd.read_table(f, **kwargs)
+        mid, ids2 = set(metadata.index), set(ids)
+        diff = mid - ids2
+        if diff:
+            logger.warning('These have metadata but do not have data - dropped: %r' % diff)
+        diff = ids2 - mid
+        if diff:
+            logger.warning('These have data but do not have metadata: %r' % diff)
+        # reorder the id in metadata to align with biom
+        metadata = metadata.loc[ids, ]
+    else:
+        metadata = pd.DataFrame(index=ids)
+    return metadata
+
+
 def read(data_file, sample_metadata_file=None, feature_metadata_file=None,
-         description='', drop=False, sparse=True, data_file_type='biom', encoding=None,
+         description='', sparse=True, data_file_type='biom',
+         sample_metadata_kwargs=None, feature_metadata_kwargs=None,
```

Comment:
what about encoding? sometimes i handle mapping files that fail with utf (so i load them with encoding='latin8'. It is really useful.